### PR TITLE
opal/accelerator: Remove function table and dlopen logic

### PR DESCRIPTION
**Pull request by wckzhang**
Instead of dlopening cuda, add direct dependency on libcuda. This also means we can remove the dlopen dependency.

Signed-off-by: William Zhang <wilzhang@amazon.com>

### CODE REVIEW

**Comment by wckzhang** [comment]
Alongside this, I want to add link libcuda against the gpusm, rgpusm, and smcuda components and directly call cuda functions within them.

File: configure.ac [hunk 1]
```diff
@@ -987,7 +987,13 @@ AC_CACHE_SAVE
 
 opal_show_title "System-specific tests"
 
-OPAL_CHECK_CUDA
+################
+# CUDA support #
```
Comment (bwbarrett) [inline configure.ac:991]: Add a note that this should be removed when the opal/cuda directory goes away.

File: opal/mca/accelerator/cuda/configure.m4 [hunk 2]
```diff
@@ -15,11 +15,12 @@
...
-#
+# This code checks makes sure the check was done earlier by the
+# opal_check_cuda.m4 code. It also copies the flags and libs under
+# opal_cuda_CPPFLAGS, opal_cuda_LDFLAGS, and opal_cuda_LIBS
 
 AC_DEFUN([MCA_opal_accelerator_cuda_CONFIG],[
+
     AC_CONFIG_FILES([opal/mca/accelerator/cuda/Makefile])
 
     # make sure that CUDA-aware checks have been done
```
Comment (bwbarrett) [inline opal/mca/accelerator/cuda/configure.m4:26]: you're a little indecisive on the interface for `OPAL_CHECK_CUDA`.  It's bad form to `AC_REQUIRE` a test that takes arguments.  You either need to update all the places that `AC_REQUIRE([OPAL_CHECK_CUDA])` to explicitly call `OPAL_CHECK_CUDA` with arguments, or have `OPAL_CHECK_CUDA` not take arguments.  My preference would be to explicitly call `OPAL_CHECK_CUDA([accelerator_cuda])` here, and then you don't need lines 33-35 below.

File: opal/mca/btl/smcuda/configure.m4 [hunk 3]
```diff
@@ -27,4 +27,13 @@ AC_DEFUN([MCA_opal_btl_smcuda_CONFIG],[
           [$1
            OPAL_MCA_CHECK_DEPENDENCY([opal], [btl], [smcuda], [opal], [common], [sm])],
           [$2])
+
+    btl_smcuda_CPPFLAGS=$opal_cuda_CPPFLAGS
```
Comment (bwbarrett) [inline opal/mca/btl/smcuda/configure.m4:31]: See comment from accelerator/cuda.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

**Comment by bwbarrett** [comment]
Overall, nice job on a big AC change.  Couple of minor things and you should be good to go.

File: opal/mca/accelerator/cuda/configure.m4 (hunk 2 above)
Comment (bwbarrett) [reply opal/mca/accelerator/cuda/configure.m4:26]: Every test in the `OPAL_CHECK_CUDA` call should be cached between runs, so it should be very quick.  If you look, this is the approach taken by UCX or OFI or similar.

Comment (wckzhang) [reply opal/mca/accelerator/cuda/configure.m4]: ACK, I changed it

**Review by bwbarrett — APPROVED** [review]

**Comment by jsquyres** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
bot:aws:retest

**Comment by wckzhang** [comment]
I'm also going to add the code that replaces opal/cuda and remove opal/cuda in this PR

**Comment by wckzhang** [comment]
Real failure when I added the opal_cuda.h header, need to figure out what I did wrong

**Comment by wckzhang** [comment]
Ah I forgot to add the opal_cuda.h file to git, should work now

**Pull request merged**